Kernel/SanCov.cpp

```diff
@@ -20,6 +38,24 @@ extern "C" void __sanitizer_cov_trace_pc(void)
         return;
 
     auto* thread = Processor::current_thread();
+
+#ifdef ENABLE_KERNEL_COVERAGE_COLLECTION_DEBUG
+    if (kcov_emergency_off) [[unlikely]]
+        return;
+
+    // Use are_interrupts_enabled() as a proxy to check we are not currently in an interrupt.
```

Comment:
But why? Can't you use `Processor::current_in_irq()`?

Reply:
See the explanation in line 64 for details. Not sure that I should duplicate the comment twice in the same file. Any suggestions how I could rephrase this?

Reply:
I understand that you do not want to collect interrupts. I do not understand why you are using `are_interrupts_enabled()` as a proxy? You comment says `..check we are not currently in an interrupt.` which I naively would implement via `if (current_in_irq()) { ..`.

So, why the proxy?

Reply:
Ahh! Yes, fair point. I don't think it makes a functional difference, but `current_in_irq()` might actually be ever so slightly faster, as it boils down to `__builtin_offsetof(Processor, m_in_irq)` which can probably be cached. `are_interrupts_enabled()` gets the cpu flags via a piece of volatile assembly, which will be re-run every time.

Will look into this. Thanks for spotting.

Reply:
Hm, actually, it turns out I already investigated this before and determined that `are_interrupts_enabled()` is better. If I swap them, serenity crashes here:
```
pwndbg> backtrace
#0  0x00000020017a226c in Kernel::halt_this () at ././Kernel/Arch/x86_64/ASM_wrapper.h:150
#1  Kernel::ProcessorBase<Kernel::Processor>::halt () at ./Kernel/Arch/x86_64/Processor.cpp:1310
#2  0x0000002000f935e6 in Kernel::__panic () at ./Kernel/Library/Panic.cpp:58
#3  0x00000020017c470d in crash_and_burn () at ./Kernel/SanCov.cpp:28
#4  __sanitizer_cov_trace_pc () at ./Kernel/SanCov.cpp:54
#5  0x000000200022006d in Kernel::read_gs_ptr () at ././Kernel/Arch/x86_64/ASM_wrapper.h:58
#6  Kernel::ProcessorBase<Kernel::Processor>::current_thread () at ././Kernel/Arch/x86_64/Processor.h:167
#7  Kernel::ProcessorBase<Kernel::Processor>::enter_trap () at ./Kernel/Arch/Processor.cpp:59
#8  0x0000002000222679 in enter_trap () at ./Kernel/Arch/TrapFrame.cpp:23
#9  0x000000200174ac36 in interrupt_common_asm_entry ()
#10 0x0000000000000000 in ?? ()
```

So we call `Processor::current_in_irq()` and somehow this results in recursion? Well yes! Check out the implementation:

https://github.com/SerenityOS/serenity/blob/ccd16c847e0283177e7bdbac2c9badc12b8bc96c/Kernel/Arch/Processor.cpp#L49-L58

`are_interrupts_enabled()` will only start returning true, once we are past `m_in_irq++`, even though we are of course already in an interrupt before then. However, before that point, we can only know this by checking if we are in a critical section via `are_interrupts_enabled()`.

Ironically, the `VERIFY_INTERRUPTS_DISABLED` will do `are_interrupts_enabled()` 🤡. So even if this did work, we would pay the same cost of crossing that volatile barrier, but run a bunch of code not knowing, that we are in an interrupt.

Reply:
I'll improve the comment.

Reply:
Extended comment.